### PR TITLE
defaults: move session persistent data to $XDG_STATE_HOME

### PR DESCRIPTION
**Pull request by etircopyh**
Since we now have [$XDG_STATE_HOME](https://gitlab.freedesktop.org/xdg/xdg-specs/-/commit/4f2884e16db35f2962d9b64312917c81be5cb54b) to stash our session state data.

### CODE REVIEW

File: runtime/doc/options.txt
Comment (justinmk) [inline runtime/doc/options.txt:827]: The `.,` prefix should remain. https://github.com/neovim/neovim/pull/15433#issuecomment-907200804

**Review by justinmk — CHANGES_REQUESTED** [review]
We can add `stdpath('state')` I guess. But moving all of backupdir, swapfiles, etc to XDG "state" location feels arbitrary and not worth the additional churn, having already changed the location in 0.5.

The distinction feels pedantic. What is the use-case? What harm is there in storing this data in the "data" dir?

**Comment by etircopyh** [comment]
> The distinction feels pedantic. What is the use-case? What harm is there in storing this data in the "data" dir?

It seems obvious. Like if you want kinda take your session to your other devices/setups. All it takes is too to copy that ~/.local/state/nvim folder or whatever your set your $XDG_STATE_HOME to.

Anyway I think sooner or later change like this will be made here as well as in other software. So it will bring some more portability and more importantly distinction, because a lot of apps can't agree on where to put this data and it ends up all over the place.

**Comment by clason** [comment]
That still doesn't explain why you can't do the same with `.local/share/nvim` -- isn't this exactly what `stdpath('data')` is for? Why do we need a _fourth_ stdpath directory?

**Comment by gpanders** [comment]
In fairness, `XDG_STATE_HOME` does sound like the correct place for many of the files currently under `XDG_DATA_HOME`:

> The `$XDG_STATE_HOME` contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in `$XDG_DATA_HOME`. It may contain:
> * actions history (logs, history, recently used files, …)
> * current state of the application that can be reused on a restart (view, layout, open files, undo history, …)

The explanation of `XDG_DATA_HOME` is much more vague:

>  `$XDG_DATA_HOME` defines the base directory relative to which user specific data files should be stored.

"User specific data files" can of course be interpreted to apply to everything listed under `XDG_STATE_HOME` as well.

I do think Nvim should adhere to defined standards, but even more importantly we should follow the principle of least surprise. Will users know or expect to look under `XDG_STATE_HOME` (I personally had never heard of it until this PR, but I'm likely not a representative sample)?

**Comment by etircopyh** [comment]
> Will users know or expect to look under XDG_STATE_HOME?

I think users that do not trouble themselves with directories like ~/.cache ~/.config ~/.local/blah-blah won't even notice the difference as they probably don't even have these directories shown in their File Manager (or whatever they using). As for more curious users I'm sure they know their way around system. And it's not like we move user configuration, session data not even supposed to be touched in most cases, let alone pushed to dotfiles repo or whatever.

> I personally had never heard of it until this PR

I'm myself didn't know about it too, until [pipewire/pipewire/commit 55c264d2](https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/55c264d2a5a776545ecd69ad7b843e6e0d795ef7). I was actually surprised as it is a good addition which brings distinction and portability.

**Comment by clason** [comment]
> I think users that do not trouble themselves with directories like ~/.cache ~/.config ~/.local/blah-blah won't even notice the difference as they probably don't even have these directories shown in their File Manager (or whatever they using). As for more curious users I'm sure they know their way around system. 

I don't think that's a safe assumption -- I for one would consider myself as someone who knows their way around my system, but having neovim files suddenly split over more directories would definitely confuse me.

> And it's not like we move user configuration, session data not even supposed to be touched in most cases, let alone pushed to dotfiles repo or whatever.

Yes, but there's possibly other data that is not synced but should be preserved. It's also a question about the plugin ecosystem -- for example, `telescope-frecency` stores the MRU database there. Is this also `state`, or should it remain in `data`?

**Comment by clason** [comment]
> But we're not splitting something important or related to data that supposed to be in $XDG_DATA_HOME, are we? AFAIK session data like *.swp files is not even kept after you successfully save a file. For `undodir` content to be persistent you have to set `undofile` as it is disabled by default. So it mostly separates files that nobody would consider editing nor touching, unless for some specific cases. For user that will want to mess with somethings in $XDG_DATA_HOME it's just a clutter.

My point is not to be too quick to assume what users want...

> As I already said all of this data is temporal by default.

Then why isn't it in `stdpath('cache')`?

> Plugin specific settings and data should be handled by plugin itself, amirite? https://github.com/nvim-telescope/telescope-frecency.nvim#configuration If to change particular plugin defaults it is to be done in plugin's upstream. There's not much n/vim can do about it.

Plugins (should) follow Neovim's conventions. If we change, they (have to) change, otherwise it'll be a mess.
 
> Also, it can be argued whether to put database file to 'data' or to 'state' IMO. But that's slightly off-topic.

No, that is exactly the point here: adding this distinction will create confusion, so there should be a clear and sufficient benefit to offset this. And I'm not seeing it yet.

**Comment by etircopyh** [comment]
> My point is not to be too quick to assume what users want...

I didn't assume what users want because it's not even possible here. It's as if users cared whether some specific code lies in file filename1.c or in filename2.c. These file is not important enough for users to care. They just logical parts of neovim functional.

> Then why isn't it in stdpath('cache')?

$XDG_CACHE_HOME (~/.cache by default) assumed to contain data that is useless enough to delete (cache in other words) at any moment. I saw some people even set XDG_CACHE_HOME to /tmp which is not persistent (like neovim session data can be and should be, sometimes). I'm not sure we're on the same page since you ask questions like this.

> Plugins (should) follow Neovim's conventions. If we change, they (have to) change, otherwise it'll be a mess.

It always was a mess, not only for n\vim but for any software with plugin system that's not being supervised by parenting project.

> No, that is exactly the point here: adding this distinction will create confusion, so there should be a clear and sufficient benefit to offset this. And I'm not seeing it yet.

What I meant is particular database should be session specific to be placed into XDG_STATE_HOME. Meaning not just any database. @gpanders provided pretty clear description right from the specs, doesn't seem to be confusing at all.

**Comment by gpanders** [comment]
To go back to @justinmk's original question:

> The distinction feels pedantic. What is the use-case? What harm is there in storing this data in the "data" dir?

Given that `XDG_STATE_HOME` is now part of the XDG specification (and assuming Neovim wants to adhere to common standards), it seems likely that we will *eventually* want to move this data to `XDG_STATE_HOME` anyway (at some point users will associate "state data" with `XDG_STATE_HOME` and be surprised that Neovim doesn't use it). So the churn either happens now and we just get it over with, or it happens later.

This also seems especially low-risk given that the files that are being moved are very rarely viewed manually.

I also don't see any problem with having Neovim state data (e.g. swap file, backups, undo history, etc.) in one location and plugin data in another (e.g. `XDG_DATA_HOME`). Plugins can use either directory (or any arbitrary directory they want, really) and it won't have any impact on Neovim.

**Comment by justinmk** [comment]
> I also don't see any problem with having Neovim state data (e.g. swap file, backups, undo history, etc.) in one location and plugin data in another (e.g. `XDG_DATA_HOME`). Plugins can use either directory (or any arbitrary directory they want, really) and it won't have any impact on Neovim.

This means in practice users would need to carry both dirs anyways. So we need to give clear guidance to plugin authors.

File: runtime/doc/vim_diff.txt
Comment (gpanders) [inline runtime/doc/vim_diff.txt:20]: No need to put "current working directory" in ``.

In fact I'm not sure this line is necessary at all. Vim doesn't place persistent session information in the current working directory (at least, not in any ways that Nvim doesn't also do).

**Review by gpanders — CHANGES_REQUESTED** [review]

**Comment by ruifm** [comment]
> But we should wait a bit for user feedback.

Random user here. I support this change and the reasons are:

* I like that my tools follow well recognized specs as close as possible. I get that `XDG_STATE_HOME` is kind of recent but such a change would take a while to reach a release.
m
* The more separation the better. I don't like logs, backups,swap files and others  hanging around the place where I have my plugins, it makes zero sense to me. If I need to backup my plugins, I would inevitably backup the logs and other ephemeral non essential files.
* I don't see how this will upset users. Beginners will not care where those files are, they never noticed them and experienced users will end up reading the docs/changelog/breaking changes anyway specially if they move to a new release.

**Comment by mjlbach** [comment]
> I don't like logs, backups,swap files and others 

Logs currently do not live in XDG_DATA_HOME because I moved them (controversially, which may get reverted) to XDG_CACHE_DIR before XDG_STATE_HOME was a thing.

**Comment by justinmk** [comment]
> The more separation the better.

Well, except that the XDG dir spec emerged to solve the problem of _too much_ separation :)

> If we do this, logs should also be moved.

Hmm, really? The use-case presented in https://github.com/neovim/neovim/pull/15583#issuecomment-913900538 is carrying a session around, without the "data" or "cache". btw, _wtf_ goes in XDG_DATA_HOME then?

From https://github.com/neovim/neovim/pull/15583#issuecomment-914347792

> The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME.

@etircopyh that description contradicts the "obvious" use-case you described in https://github.com/neovim/neovim/pull/15583#issuecomment-913900538 .  "Take a session to other devices/setups" implies _portal_ state.

- undofiles are _not_ portable, they assume a particular filesystem layout
- swapfiles are _not_ portable for the same reason
- backupdirs are _not_ portable for the same reason

**Comment by mjlbach** [comment]
@justinmk I was just going off the description in the XDG spec which lists logging/MRU info

```
    The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME. It may contain:

    actions history (logs, history, recently used files, …)

    current state of the application that can be reused on a restart (view, layout, open files, undo history, …)
````

**Comment by gpanders** [comment]
Windows CI is failing functionaltests:

```
[  ERROR   ] test/functional\options\defaults_spec.lua @ 617: stdpath() returns a String with "state" doesn't expand ~/
test\helpers.lua:73: Expected objects to be the same.
Passed in:
(string) '~\frobnitz\nvim'
Expected:
(string) '~\frobnitz\nvim-data'
```

You'll also need to modify your commit messages to pass the commit linter. Commits that update or modify the tests should begin with `test:` and the other commits should begin with `feat:` (indicating a new feature, i.e. support for `XDG_STATE_HOME`).

**Comment by clason** [comment]
Now we are moving logs around, too (again)...

Can we at least add a `stdpath('log')` (even if it's not a standard XDG directory -- although it should be!) and use that so we can have a consistent and transparent location that plugins can rely on, too? (That also makes it easier to change the location once more later on when XDG introduces new directories...)

**Comment by etircopyh** [comment]
> Well, except that the XDG dir spec emerged to solve the problem of too much separation :)

It doesn't cause "too much separation" if the content being separated is mostly a junk to the end user.

> - undofiles are not portable, they assume a particular filesystem layout
> - swapfiles are not portable for the same reason
> - backupdirs are not portable for the same reason

I remember reading about shada file being specific to particular user and being somewhat restricted in access rights due to security reasons. Not sure what you mean by undo, swap and backup "assuming a particular filesystem layout", I guess I'll have to read up on that. Anyways, I presented a "use case" (basically the first thing that popped into my head) because you asked me to. I don't think XDG Base Dirs spec need a "use" for a case, it's just a set of directories to provide a standard.

**Comment by clason** [comment]
> I don't think XDG Base Dirs spec need a "use" for a case, it's just a set of directories to provide a standard.

Adding support for the new base dir is uncontroversial (so this is a straw man argument); the controversial part is moving Neovim's user files around.

> It doesn't cause "too much separation" if the content being separated is mostly a junk to the end user.

Again, one person's junk is another person's treasure.

**Comment by etircopyh** [comment]
> Can we at least add a stdpath('log') (even if it's not a standard XDG directory -- although it should be!)
Wait, what? It's been referenced like a few times already that $XDG_STATE_HOME is meant to hold stuff like logs, app session state and other non-essential to user but possibly persistent files.

First you speak against the separation and now you want another dir just for logs. I a little confused now.

**Comment by mjlbach** [comment]
My understanding is clason just wants `stdpath('log')` as an alias to the on disk location (which we talked about last time I moved the log directory) such that the change referenced in this PR would be transparent to plugin developers/users

**Comment by clason** [comment]
I don't mean a new directory, I mean having an unambiguous standard directory that plugins can access for their own logs. This could point to any of the existing (or newly introduced) XDG directories.

The point is that nobody needs to remember whether logs now live in `XDG_CACHE`, `XDG_DATA`, or `XDG_STATE` and only need write to (or look in) `stdpath('log')` from within neovim.

Changing the location later is then just a single change to the definition of `stdpath('log')`, so _all_ logs move at the same time.

**Comment by etircopyh** [comment]
Ok, got it. Thanks for clarification. Yes, it makes sense for neovim to make some standardization of its own :)

**Comment by etircopyh** [comment]
> Can we at least add a `stdpath('log')`

Done

**Comment by etircopyh** [comment]
It's good to merge, I guess.

File: src/nvim/os/stdpaths.c
Comment (zeertzjq) [inline src/nvim/os/stdpaths.c:156]:
```suggestion
                                  const bool escape_commas)
```

**Comment by etircopyh** [comment]
So, will this PR make its way into 0.6.2 (0.7?). Does anyone have any remarks or objections?

**Comment by etircopyh** [comment]
> So, will this PR make its way into 0.6.2 (0.7?). Does anyone have any remarks or objections?

🗿

**Comment by mjlbach** [comment]
I'm sorry your PR has been neglected, unfortunately we don't have very many contributors with commit access, and the last time we changed the location to cache it was fairly controversial (before the XDG spec for state home was finalized). @justinmk is probably going to have to give the go-ahead to this before I or anyone would be able to merge this because he had some concerns when I moved it last time.

**Review by justinmk — APPROVED** [review]
LGTM, might need a rebase for a final check

**Comment by etircopyh** [comment]
> LGTM, might need a rebase for a final check

Done

**Comment by etircopyh** [comment]
🤔 What's up with windows tests?

File: runtime/doc/options.txt
Comment (dundargoc) [inline runtime/doc/options.txt:3505]:
```suggestion
	expand "$HOME/.local/state/nvim/shada/main.shada".  Maybe you should
```

File: runtime/doc/starting.txt
Comment (dundargoc) [inline runtime/doc/starting.txt:1357]:
```suggestion
By default, the file is located at stdpath('log')/log unless that path
```

File: runtime/doc/starting.txt
Comment (dundargoc) [inline runtime/doc/starting.txt:1332]:
```suggestion
The $XDG_CONFIG_HOME, $XDG_DATA_HOME and $XDG_STATE_HOME environment variables
are used if they exist, otherwise default values (listed below) are used.
```

**Comment by dundargoc** [comment]
> thinking What's up with windows tests?

```
Passed in:
(string) '.,$XDG_CONFIG_HOME/nvim/backup//'
Expected:
(string) '.,$XDG_CONFIG_HOME/nvim-data/backup//'
```

It wants `nvim-data` for some reason?

**Comment by zeertzjq** [comment]
Does this require users to move their files manually after updating Nvim?

**Comment by justinmk** [comment]
> It wants `nvim-data` for some reason?

nvim-data is the compromise on Windows because it doesn't have XDG-style directories. 

I'd suggest keeping the name `nvim-data` (i.e. don't rename it to `nvim-state`) for the same reason: this PR has little relevance to Windows (except for the env var names).

**Comment by etircopyh** [comment]
> It wants `nvim-data` for some reason?

Yeah, I got that. I just don't get why it can't set the right name (nvim-data) for Windows. It's like iswin() function returns false for some reason 🤔

**Review by justinmk — APPROVED** [review]
Looks like `:help stdpath()` needs an update to mention "state" and "log"

File: runtime/doc/builtin.txt
Comment (justinmk) [inline runtime/doc/builtin.txt:7478]: I rearranged this to maintain alphabetical order.

**Comment by justinmk** [comment]
@etircopyh thanks for your persistence here, it increases trust in the result. 

I squashed the commits because they were over-granular and I think this fits as a single unit.

**Comment by dundargoc** [comment]
The windows jobs are failing.

**Comment by DusanLesan** [comment]
@etircopyh I lost a lot of time debugging my configs to see why nvim started making a new directory in my .local
I do not use XDG_STATE_HOME so I just set it as $HOME/.local/share but can I set storage location just for nvim in its configs?

**Comment by justinmk** [comment]
> but can I set storage location just for nvim in its configs?

Yes, by setting the various XDG_ env vars documented at `:help standard-path`.

**Comment by sisrfeng** [comment]
> > but can I set storage location just for nvim in its configs?
> 
> Yes, by setting the various XDG_ env vars documented at `:help standard-path`.

This answer suggests users to set XDG_STATE_HOME, but what @DusanLesan   wants is something  only affect nvim, right?

**Comment by sisrfeng** [comment]
> @etircopyh I lost a lot of time debugging my configs to see why nvim started making a new directory in my .local I do not use XDG_STATE_HOME so I just set it as $HOME/.local/share but can I set storage location just for nvim in its configs?

This lesson tells  us the necessity of following this issue https://github.com/neovim/neovim/issues/14090.
(I can not find the message in the release not of nvim 0.8)

**Pull request merged**
